org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -2165,17 +2165,20 @@ private TypeVariableBinding[] getTPrimeArgumentsAndCreateBounds(
 			InferenceVariable[] beta) {
 		TypeBinding[] arguments = parameterizedType.typeArguments();
 		TypeVariableBinding[] typeVariables = new TypeVariableBinding[arguments.length];
+		InferenceSubstitution theta = new InferenceSubstitution(this.environment, beta, this.currentInvocation);
```

Comment:
Here and usage of theta below: update from the spec where it says: "_let θ be the substitution[Z1:=β1, ..., Zk:=βk]. T' is Tc θ._" etc.

Note, that we don't seem to have any tests that actually **need** this.

Reply:
> Here and usage of theta below: update from the spec where it says: "_let θ be the substitution[Z1:=β1, ..., Zk:=βk]. T' is Tc θ._" etc.
> 
> Note, that we don't seem to have any tests that actually **need** this.

`sealed interface I<T> {}
sealed interface J<T> {}
class A  {}
record R<T>(T t) implements I<T>, J<T>{}

public class X<T extends I<? extends A> & J<? extends A>> {

    public boolean foo(T t) {
    	return switch(t) {
    		case R(A a) -> true;
    		default -> false;
    	};
    }

    public static void main(String argv[]) {
       System.out.println(new X<R<? extends A>>().foo(new R(new A())));
    }
}`
the code above actually hits that line where theta is created.. and while javac passes ecj gives an error "Type T cannot be safely cast to R<? extends A>"

Reply:
I assume `I` and `J` should be generic (this time it wasn't github swallowing angle brackets).

But then I get
```
3. ERROR in X.java (at line 16)
	System.out.println(new X<R<? extends A>>().foo(new R(new A())));
	                         ^
Bound mismatch: The type R<? extends A> is not a valid substitute for the bounded parameter <T extends I<? extends A> & J<? extends A>> of the type X<T>
```

How should I complete the example?

Reply:
> I assume `I` and `J` should be generic (this time it wasn't github swallowing angle brackets).

yes, it should be generic - I don't know how this gobbling up happened again - I've corrected the same.  Could you please check whether you are getting the error? I am not seeing this error - noting that I've applied your patch and then checking, just to be sure we are on the same page.

Reply:
Thanks, now I got past those unintended errors. While I was at it, I took the liberty to use unique names for all type parameters in the test, because working with type `T` is quite cumbersome, if *every* type parameter is called `T` :)

Reply:
Test included in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1341/commits/0ace862918733f41e5f74520152b0575f79df142, but still I don't think we really exercised the substitution stuff ...